### PR TITLE
feat: added near-mock-vm and reorganize into mono-repo

### PR DESCRIPTION
**Pull request by willemneal**
Previously `near-mock-vm` lived in a open PR of `near-sdk-rs`.  That PR now just handles the changes needed to the rust sdk to make the crate possible.  This PR also reorganizes the repo into a mono-repo.  While the sub packages will be published end users will still only use `near-sdk-as`.

Update near-mock-vm so that there are no gas or storage fees charged.  This way unit tests won't fail because of gas limits.  Also added missing API

### CODE REVIEW

File: near-mock-vm/src/mocked_chain.rs [hunk 1]
```diff
@@ -0,0 +1,1253 @@
+use serde::Serialize;
+
+use crate::memory::*;
+use crate::utils::*;
+use crate::runner;
+use near_sdk::*;
+use near_vm_logic::MemoryLike;
+use std::collections::HashMap;
+use wasm_bindgen::prelude::*;
```
Comment (frol) [inline near-mock-vm/src/mocked_chain.rs:9]: It is quite hard to reason about the code with `*` imports.

File: near-mock-vm/src/mocked_chain.rs [hunk 2]
```diff
@@ -0,0 +1,1253 @@
...
+    /// `base + read_register_base + read_register_byte * num_bytes + write_memory_base + write_memory_byte * num_bytes`
+    pub fn read_register(&mut self, register_id: u64, ptr: u64) {
+        // let data = &vec![42];
+        //self.chain.wrapped_internal_write_register(register_id, &data);
+        unsafe { self.chain.read_register(register_id, ptr) }
+        // match res {
+        //     Ok(()) => (),
+        //     // Err(VMLogicError::HostError(_e)) => console_log!("Host Error"),
+        //     Err(e) => panic!(e)
+        // }
```
Comment (frol) [inline near-mock-vm/src/mocked_chain.rs:207]: Should this be cleaned up? Is it WIP?

File: near-mock-vm/vm/tests/utils.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+
```
Comment (frol) [inline near-mock-vm/vm/tests/utils.js:1]: Is there a way to reuse `near-mock-vm/vm/src/utils.ts` instead of copy-pasting?

File: near-mock-vm/vm/tsconfig.json [hunk 4]
```diff
@@ -2,7 +2,7 @@
   "compilerOptions": {
     /* Basic Options */
     "incremental": true,                      /* Enable incremental compilation */
-    "target": "ESNEXT",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "target": "ESNext",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```
Comment (frol) [inline near-mock-vm/vm/tsconfig.json:5]: Is this change necessary?

File: sdk/package.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "near-sdk-as",
+  "version": "2.0.0-next.1",
+  "description": "Collection of packages for developming NEAR smart contracts",
```
Comment (frol) [inline sdk/package.json:4]:
```suggestion
  "description": "Collection of packages for developing NEAR smart contracts in AssemblyScript",
```

**Review by frol — APPROVED** [review]
Should `.wat` files (` near-mock-vm/pkg/near_mock_vm_bg.wat`) be gitignored?

I have just scanned through the PR (it is massive!), overall it looks good to me

File: bindgen/package.json [hunk 6]
```diff
@@ -1,9 +1,9 @@
 {
-  "name": "bindgen",
-  "version": "0.0.1-alpha.0",
+  "name": "near-sdk-bindgen",
+  "version": "2.0.0-next.1",
```
[the remark this answers is not in the document]
Comment (willemneal) [reply bindgen/package.json:3]: Before this package was never published.  The package.json was just used for the testing/build scripts.  Now they should also have the same version.

File: near-mock-vm/vm/tests/utils.js (hunk 3 above)
Comment (willemneal) [reply near-mock-vm/vm/tests/utils.js:1]: Yeah this was an old artifact.

File: near-mock-vm/vm/tsconfig.json (hunk 4 above)
Comment (willemneal) [reply near-mock-vm/vm/tsconfig.json:5]: I'm not sure how this happened; I just moved the file.  I'll revert it though.

File: sdk-core/package.json [hunk 7]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "near-sdk-core",
+  "version": "2.0.0-next.1",
+  "description": "Core AssemblyScript package for developming NEAR smart contracts",
```
Comment (willemneal) [inline sdk-core/package.json:4]:
```suggestion
  "description": "Core AssemblyScript package for developing NEAR smart contracts",
```

File: README.md [hunk 8]
```diff
@@ -1,18 +1,54 @@
 # near-sdk-as
+
 Collection of packages used in developing NEAR smart contracts in AssemblyScript including:
 
 
-- [`runtime`](./assembly/runtime) - AssemblyScript near runtime library
+- [`sdk`](./sdk) - AssemblyScript near runtime library
```
[the remark this answers is not in the document]
Comment (willemneal) [reply README.md:6]: Well the term `runtime` is a bit overloaded as in the simulator we create a mock runtime.  I was mainly trying to follow the rust sdk's naming.  Though it should point to sdk-core.

File: near-mock-vm/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+# all the `std::fmt` and `std::panicking` infrastructure, so isn't great for
+# code size when deploying.
+console_error_panic_hook = { version = "0.1.1", optional = true }
+
+# `wee_alloc` is a tiny allocator for wasm that is only ~1K in code size
+# compared to the default allocator's ~10K. It is slower than the default
+# allocator, however.
+#
+# Unfortunately, `wee_alloc` requires nightly Rust when targeting wasm for now.
+#wee_alloc = { version = "0.4.2", optional = true }
```
[the remark this answers is not in the document]
Comment (willemneal) [reply near-mock-vm/Cargo.toml:39]: I just turned it into a feature.  I don't think that people would want it one unless it was going to be used in a browser context, since it's faster without it.

**Comment by willemneal** [comment]
Thanks @frol!
